From f8449c05fd77cd834ab0f48bd0c8351d61d42848 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Thu, 14 Aug 2014 14:42:16 -0400
Subject: [ethernet] e1000e: Fix issue with link flap on 82579

Message-id: <1408027348-9610-23-git-send-email-jogreene@redhat.com>
Patchwork-id: 87716
O-Subject: [RHEL7.1 PATCH 22/34] e1000e: Fix issue with link flap on 82579
Bugzilla: 1091119
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Several customers have reported a link flap issue on 82579. The symptoms
are random and intermittent link losses when 82579 is connected to specific
link partners. Issue has been root caused as interoperability problem
between 82579 and at least some Broadcom PHYs in the Energy Efficient
Ethernet wake mechanism.

To fix the issue, we are disabling the Phase Locked Loop shutdown in 100M
Low Power Idle.  This solution will cause an increase of power in 100M EEE
link. It will cost additional 28mW in this specific mode.

Cc: Lukasz Adamczuk <lukasz.adamczuk@intel.com>
Signed-off-by: Dave Ertman <davidx.m.ertman@intel.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
(cherry picked from commit 7142a55c3c1fee89a60aa7b402c834b6b8afcb0a)

BZ: 1091119
Testing: Acceptable, see BZ
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7834849
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/e1000e/ich8lan.c b/drivers/net/ethernet/intel/e1000e/ich8lan.c
index 77e7964..80806bd 100644
--- a/drivers/net/ethernet/intel/e1000e/ich8lan.c
+++ b/drivers/net/ethernet/intel/e1000e/ich8lan.c
@@ -814,6 +814,17 @@ s32 e1000_set_eee_pchlan(struct e1000_hw *hw)
   }
  }
 
+ if (hw->phy.type == e1000_phy_82579) {
+  ret_val = e1000_read_emi_reg_locked(hw, I82579_LPI_PLL_SHUT,
+          &data);
+  if (ret_val)
+   goto release;
+
+  data &= ~I82579_LPI_100_PLL_SHUT;
+  ret_val = e1000_write_emi_reg_locked(hw, I82579_LPI_PLL_SHUT,
+           data);
+ }
+
  /* R/Clr IEEE MMD 3.1 bits 11:10 - Tx/Rx LPI Received */
  ret_val = e1000_read_emi_reg_locked(hw, pcs_status, &data);
  if (ret_val)
diff --git a/drivers/net/ethernet/intel/e1000e/ich8lan.h b/drivers/net/ethernet/intel/e1000e/ich8lan.h
index 8fc6c15..5515126 100644
--- a/drivers/net/ethernet/intel/e1000e/ich8lan.h
+++ b/drivers/net/ethernet/intel/e1000e/ich8lan.h
@@ -232,12 +232,14 @@
 #define I82577_MSE_THRESHOLD 0x0887 /* 82577 Mean Square Error Threshold */
 #define I82579_MSE_LINK_DOWN 0x2411 /* MSE count before dropping link */
 #define I82579_RX_CONFIG  0x3412 /* Receive configuration */
+#define I82579_LPI_PLL_SHUT  0x4412 /* LPI PLL Shut Enable */
 #define I82579_EEE_PCS_STATUS  0x182E /* IEEE MMD Register 3.1 >> 8 */
 #define I82579_EEE_CAPABILITY  0x0410 /* IEEE MMD Register 3.20 */
 #define I82579_EEE_ADVERTISEMENT 0x040E /* IEEE MMD Register 7.60 */
 #define I82579_EEE_LP_ABILITY  0x040F /* IEEE MMD Register 7.61 */
 #define I82579_EEE_100_SUPPORTED (1 << 1) /* 100BaseTx EEE */
 #define I82579_EEE_1000_SUPPORTED (1 << 2) /* 1000BaseTx EEE */
+#define I82579_LPI_100_PLL_SHUT (1 << 2) /* 100M LPI PLL Shut Enabled */
 #define I217_EEE_PCS_STATUS 0x9401 /* IEEE MMD Register 3.1 */
 #define I217_EEE_CAPABILITY 0x8000 /* IEEE MMD Register 3.20 */
 #define I217_EEE_ADVERTISEMENT 0x8001 /* IEEE MMD Register 7.60 */
-- 
1.7.1